lib/dialogue-agent/conversation.ts

```diff
@@ -571,40 +597,61 @@ export default class Conversation extends events.EventEmitter {
         return this._addMessage({ type: MessageType.LINK, url, title });
     }
 
-    appendLog(turn : DialogueTurn) {
-        this._log.push(turn);
+    lastDialogue() {
+        if (this._log.length === 0)
+            throw new Error('No dialogue is logged');
+        return this._log[this._log.length - 1];
+    }
+
+    lastTurn() {
+        const dialogue = this.lastDialogue();
+        if (dialogue.turns.length === 0)
+            throw new Error('No dialogue is logged');
+        return dialogue.turns[dialogue.turns.length - 1];
+    }
+
+    appendNewDialogue() {
+        this._log.push(new DialogueLog());
+    }
+
+    appendNewTurn(turn : DialogueTurn) {
+        if (this._log.length === 0 || this.lastDialogue().done)
+            this.appendNewDialogue();
+        const dialogue = this.lastDialogue();
+        dialogue.append(turn);
     }
 
     voteLast(vote : 'up'|'down') {
-        if (this._log.length === 0)
-            throw new Error('No response found to be voted');
-        const last = this._log[this._log.length - 1];
+        const last = this.lastTurn();
         last.vote = vote;
     }
 
     commentLast(comment : string) {
-        if (this._log.length === 0)
-            throw new Error('No response found to be commented');
-        const last = this._log[this._log.length - 1];
+        const last = this.lastTurn();
         last.comment = comment;
     }
 
     async saveLog() {
-        const dir = this._engine.platform.getWritableDir();
+        const dir = path.join(this._engine.platform.getWritableDir(), 'logs');
+        if (!fs.existsSync(dir))
+            fs.mkdirSync(dir);
```

Comment:
Technically, this is bad practice: you should never check that a file exists before doing something. You should try to make the directory, and ignore the error if it's EEXIST. This avoids a race condition (also known as Time-Of-Check-To-Time-Of-Use vulnerability). Here, it's harmeless, but there is no reason to write bad code.